src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+        /// Derived2 is not a potential conversion type.
+        /// 
+        /// arguments: The arguments of invocation expression
+        /// parameters: The parameters of function
+        /// targetArgument: The target argument that contains target node
+        /// targetParamIndex: Output the corresponding parameter index of the target arugment if function returns true
+        /// <returns>
+        /// True, if arguments and parameters match perfectly.
+        /// False, otherwise.
+        /// </returns>
+        // TODO: May need an API to replace this function,
+        // link: https://github.com/dotnet/roslyn/issues/42149
+        private static bool IsArgumentListAndParameterListPerfectMatch(SemanticModel semanticModel, SeparatedSyntaxList<ArgumentSyntax> arguments,
+            ImmutableArray<IParameterSymbol> parameters, ArgumentSyntax targetArgument, CancellationToken cancellationToken, out int targetParamIndex)
+        {
+            targetParamIndex = -1; // return invalid index if it is not a perfect match
+
+            var matchedTypes = new bool[parameters.Length]; // default value is false
+            var paramsMatchedByArray = false; // the parameter with keyword params can be either matched by an array type or a variable number of arguments
+            var inOrder = true; // assume the arguments are in order in default
+
+            for (var i = 0; i < arguments.Count; i++)
+            {
+                // Parameter index cannot out of its range, #arguments is larger than #parameter only if the last parameter with keyword params
+                var parameterIndex = Math.Min(i, parameters.Length - 1);
+
+                // If the argument has a name, get the corresponding parameter index
+                var nameSyntax = arguments[i].NameColon?.Name;
+                if (nameSyntax != null)
+                {
+                    var name = nameSyntax.ToString();
+                    var found = false;
+                    for (var j = 0; j < parameters.Length; j++)
+                    {
+                        if (name.Equals(parameters[j].Name))
+                        {
+                            // Check if the argument is in order with parameters.
+                            // If the argument breaks the order, the rest arguments of matched functions must have names
+                            if (i != j)
+                                inOrder = false;
+                            parameterIndex = j;
+                            found = true;
+                            break;
+                        }
+                    }
+                    if (!found) return false;
+                }
+
+                // The argument is either in order with parameters, or have a matched name with parameters
+                var argType = semanticModel.GetTypeInfo(arguments[i].Expression, cancellationToken);
+                if (argType.Type != null && (inOrder || nameSyntax is object))
```

Comment:
my brain is literally too weak to read and understand this.

Reply:
I add all rules for the matching, and comment them on the code that might be confusing

https://github.com/dotnet/roslyn/blob/003a8d9cfda7c1ecf9a2e0b9a98dcfa6af12d9be/src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs#L263-L275